### DISCUSSION TITLE
JIB Extension and docker layers

### DISCUSSION THREAD

**Discussion by fbotis** [quarkusio/quarkus#26731] [Q&A] [answered]
I'm using jib extension to generate the docker container but I noticed that even if I just do a application.properties code change (basically changing only the "app" layer) the digest of the lib layer changes as well so when I do a docker pull it has to download all the "lib" layer again (which is the bigger part of the app)

See below screenshot (above the first version of the container and below the one where I just did a config change in application.properties)
<img width="943" alt="Screenshot 2022-07-14 at 11 17 16" src="https://user-images.githubusercontent.com/1549438/178948567-605423fe-1d08-4e66-8bd8-7333fc07ceb5.png">

I suspect this might be related to the ArcProcessor 

Do you guys know if there's any workaround this so the "lib" layer doesn't change if the content doesn't either?

[a comment by an automated account is not shown]

**Comment by geoand** [comment, marked as answer]
Changing `application.properties` could actually result in very different Quarkus behavior as you could be changing build time properties. 

I wish we had a separation between `application-runtime.properties` and `application-buildtime.properties` which would allow us to make such a distinction, but we don't...

**Reply by fbotis** [reply]
Thanks for the quick reply!

**Reply by geoand** [reply]
YW

**Reply by fbotis** [reply]
The thing is I get the same behavior even if I don't do any changes in the code/config, I just build the app twice and the digest on the "lib" layer is different each time. Is that the normal behavior?

**Reply by geoand** [reply]
Yes, when the application is rebuilt, we don't know what changed

**Reply by fbotis** [reply]
Do you think setting the timestamp of the files to a constant would fix that problem , in io.quarkus.container.image.jib.deployment.JibProcessor#addLayer

`addLayer(jibContainerBuilder, nonFastChangingLibPaths, libsMainPath, "fast-jar-normal-libs",
                        isMutableJar, now);`

Instead of **now** use EPOCH_TIMESTAMP for example?

**Reply by geoand** [reply]
You can certainly try it out :)

**Reply by fbotis** [reply]
I think there's  no point of using jib at all if the behavior is to change the lib layer each time...I think also the documentation is wrong (from https://quarkus.io/guides/container-image): 

_The extension quarkus-container-image-jib is powered by [Jib](https://github.com/GoogleContainerTools/jib) for performing container image builds. **The major benefit of using Jib with Quarkus is that all the dependencies (everything found under target/lib) are cached in a different layer than the actual application making rebuilds really fast** and small (when it comes to pushing). Another important benefit of using this extension is that it provides the ability to create a container image without having to have any dedicated client side tooling (like Docker) or running daemon processes (like the Docker daemon) when all that is needed is the ability to push to a container image registry._

Either that, either I'm using it wrong :)

**Reply by fbotis** [reply]
@geoand I updated JibProcessor to use Instant.EPOCH instead of now locally and gave it a go, looks like it solves the problem, now the "lib" layer has the same digest on multiple builds

JibProcessor.java:494

```
addLayer(jibContainerBuilder, nonFastChangingLibPaths, libsMainPath, "fast-jar-normal-libs",
                        isMutableJar, Instant.EPOCH);
                        
addLayer(jibContainerBuilder, new ArrayList<>(fastChangingLibPaths), libsMainPath, "fast-jar-changing-libs",
                        isMutableJar, Instant.EPOCH);
```
                        
Shall I create a ticket and a PR for this?

**Reply by geoand** [reply]
Yes please. We'll need to hide this change behind a configuration option (that is false by default to keep backwards compatibility)
